cibuildwheel/util.py

```diff
@@ -81,18 +82,21 @@ def download(url, dest):
         response.close()
 
 
+DependencyConstraints_T = TypeVar('DependencyConstraints_T', bound='DependencyConstraints')
+
+
 class DependencyConstraints:
-    def __init__(self, base_file_path):
+    def __init__(self, base_file_path: str):
         assert os.path.exists(base_file_path)
         self.base_file_path = os.path.abspath(base_file_path)
 
     @classmethod
-    def with_defaults(cls):
+    def with_defaults(cls: Type[DependencyConstraints_T]) -> DependencyConstraints_T:
```

Comment:
Because `cls` is a parameter and could be different if there's a subclass. If this isn't necessary, this should probably not be `@classmethod` but `@staticmethod`?

Reply:
I still does not understand. If it is subclass it still return instance of `DependecyConstrains` and cls still be `DependecyConstrains` class (with extension)

Reply:
No, because you expect `SubclassOfDependencyConstraints.with_defaults` to return a `SubclassOfDependencyConstraints` object, and not a `DependencyConstraints` object. So the return statement would be OK, but the code using the classmethod wouldn't be OK.

Reply:
Anyway, it's solved. See https://github.com/joerick/cibuildwheel/pull/315#discussion_r406854654.

Reply:
So still typing system in python have weakness. Declaration that it return something in type of current class should be easy.